### PR TITLE
Factor out the lifecycle testing framework

### PR DESCRIPTION
**Pull request by lunaris**
Lifecycle tests exist to test the Pulumi engine, allowing us to mock providers and programs and resource registrations before executing specific operations. We'd like to extend our lifecycle test suite to support fuzzing the engine -- that is, generating test cases to exercise edge cases that we might otherwise struggle to find manually. Presently, the framework underpinning the lifecycle tests is defined in a single `test_plan.go` module that sits alongside the `_test.go` files themselves. This means that it's not really possible to factor out or neatly extend the framework (e.g. by adding a `fuzzing` submodule) without introducing circular imports (since tests in `lifecycletest` will import `fuzzing`, which imports the framework from `lifecycletest`).

This commit therefore pulls out `test_plan.go` into a dedicated `framework` submodule, which is now explicitly imported and used by existing tests. It also adapts the framework slightly to define the subset of `testing.T` functionality it needs more precisely, so that when the time comes for us to fuzz we can e.g. pass a `rapid.T` just as easily. No behavioural changes are made to the tests themselves -- this is purely a structural change to facilitate subsequent work on fuzzing.

### CODE REVIEW

File: pkg/engine/lifecycletest/framework/framework.go [hunk 1]
```diff
@@ -59,6 +57,26 @@ import (
 	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
 )
 
+// TB is a subset of testing.TB that admits other T-like things, such as *rapid.T from the Rapid property-testing
+// library. It covers the set of functionality that we actually need for lifecycle testing, and satisfies the interfaces
+// of testify.assert and testify.require.
+type TB interface {
```
Comment (brandonpollack23) [inline pkg/engine/lifecycletest/framework/framework.go:63]: So I know this is the name in the framework but I want to just ask:

Are we certain this (terrible but consistent) name is worth keeping? Perhaps we should consider something more descriptive like TestingSharedFunctionality (equally bad).

They're both not great but figured I'd raise it

File: pkg/backend/httpstate/snapshot_test.go [hunk 2]
```diff
@@ -38,7 +38,7 @@ import (
 
 	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate/client"
 	"github.com/pulumi/pulumi/pkg/v3/engine"
-	"github.com/pulumi/pulumi/pkg/v3/engine/lifecycletest"
+	lt "github.com/pulumi/pulumi/pkg/v3/engine/lifecycletest/framework"
```
Comment (brandonpollack23) [inline pkg/backend/httpstate/snapshot_test.go:41]: just a nit: ltf?

**Review by brandonpollack23 — APPROVED** [review]
lgtm, just a couple naming nits that aren't necessary just opening for discussion.  If you could comment here for posterity if you did this manually or a neat refactoring strategy worked well here that'd be cool (in go I find having to capitalize the first letter of a function and prepend makes moving stuff difficult sometimes)

File: pkg/engine/lifecycletest/framework/framework.go (hunk 1 above)
Comment (lunaris) [reply pkg/engine/lifecycletest/framework/framework.go:63]: Discussed offline and I think we're going to keep this as-is since `TB` is something of a standard in the Go language itself.
